### DISCUSSION TITLE
Recovering from table schema load error

### DISCUSSION THREAD

**Discussion by ljfc** [jeremyevans/sequel#1892] [General]
Hello, we’re having occasional database connection issues that are nothing to do with Sequel, but they are causing a knock-on effect in Sequel which I don’t know how to recover from.

What happens is:

1. The table schema load fails to connect to the DB: `PG::UnableToSend: SSL SYSCALL error: EOF detected: SELECT \"pg_attribute\".\"attname\" AS \"name\", CAST(\" ... etc`
2. The _next_ query relating to that table fails because Sequel has forgotten that a column is JSON, and therefore sends inappropriate SQL: `PG::UndefinedColumn: ERROR: column \"customer_id\" does not exist`
3. The transaction rolls back

What I’d like to know is how I ought to be recovering from this? Nothing is raised when 1 fails, only when 2 does, so I don’t see an opportunity to retry at that point. If I rescue `PG::UndefinedColumn` then will `Database#schema(:billing_infos, reload: true)` reload the schema that Sequel::Model also refers to?

If I _always_ want to retry something like this at least once, is there a way to tell that to Sequel? I’m happy to structure higher-level code so it always tries to rescue and repair the situation, but maybe I’m missing something here.

**Notes**

The error in 2 is strictly correct, column `customer_id` does not exist. But that’s not what’s being asked for, as seen in the Ruby code when this error occurs:

```
invoice.user.add_billing_info({
  payment_method: "stripe",
  payment_details: {
    customer_id: payment_method.customer,
    payment_method_id: payment_method.id,
    type: payment_method.type
  }
})
```

Here `invoice` is a Sequel::Model instance, `payment_method` is an object we got from Stripe’s API, and the relations `invoice.user` and `user.billing_infos` are configured and work. Crucially `payment_details` is a JSON column, so `customer_id` is supposed to be a key in a JSON object, not a column name.

This code works as intended _most_ of the time!

**Comment by ljfc** [comment]
Maybe a slight tangent, but if I implemented schema caching, would that prevent this specific error because the schema is guaranteed to be available?

(Obviously we’d still see _some_ error when the DB connection glitches, but this particular issue was harder to track down than a ‘regular’ DB connection problem 😅 )

**Comment by jeremyevans** [comment]
> Hello, we’re having occasional database connection issues that are nothing to do with Sequel, but they are causing a knock-on effect in Sequel which I don’t know how to recover from.
> 
> What happens is:
> 
> 1. The table schema load fails to connect to the DB: `PG::UnableToSend: SSL SYSCALL error: EOF detected: SELECT \"pg_attribute\".\"attname\" AS \"name\", CAST(\" ... etc`

By default, a failure to load the model schema should raise an exception, it shouldn't be silently swallowed.  If it is being silently swallowed, my guess is you have set require_valid_table to false, which is a terrible idea.  Alternatively, you are using a very old version of Sequel and should upgrade.

> 2. The _next_ query relating to that table fails because Sequel has forgotten that a column is JSON, and therefore sends inappropriate SQL: `PG::UndefinedColumn: ERROR: column \"customer_id\" does not exist`

This would be expected if Sequel cannot determine the table schema.  Obviously, this is a bad thing, which is why Sequel sets require_valid_table to true by default.

> Maybe a slight tangent, but if I implemented schema caching, would that prevent this specific error because the schema is guaranteed to be available?

Yes, the schema_caching extension would probably help prevent this particular type of error.

One potential issue I see is that the UnableToSend exception is not treated as a disconnect error. What happens on future use of the same connection?  Do you get a different error that is treated as a disconnect error?  I would assume you must, otherwise you would have a disconnected connection in the pool that would continuously cause problems every time it is used.  You could try the following patch and see if helps, though it is unlikely to be related to this specific issue:

```diff
index fe9292c30..0eeec582a 100644
--- a/lib/sequel/adapters/postgres.rb
+++ b/lib/sequel/adapters/postgres.rb
@@ -52,6 +52,9 @@ module Sequel
       if defined?(::PG::ConnectionBad)
         DISCONNECT_ERROR_CLASSES << ::PG::ConnectionBad
       end
+      if defined?(::PG::UnableToSend)
+        DISCONNECT_ERROR_CLASSES << ::PG::UnableToSend
+      end
       DISCONNECT_ERROR_CLASSES.freeze

       disconnect_errors = [
```

**Reply by ljfc** [reply]
> By default, a failure to load the model schema should raise an exception, it shouldn't be silently swallowed. If it is being silently swallowed, my guess is you have set require_valid_table to false, which is a terrible idea. Alternatively, you are using a very old version of Sequel and should upgrade.

Thank you, yes that makes sense, I was confused that this didn’t raise. I can’t see `require_valid_table` being set to false anywhere in our codebase, so I don’t think it’s that. We’re on 5.34 – is that old enough to have this issue? In any case, we will upgrade. Otherwise I don’t see why this is getting swallowed, we’re certainly not rescuing anything in that part of our code. This _is_ all in DB transaction, but that only swallows `Sequel::Rollback` without re-raising I believe, and in any case, it’s only the _second_ error which causes the transaction to roll back.

> What happens on future use of the same connection? Do you get a different error that is treated as a disconnect error? I would assume you must, otherwise you would have a disconnected connection in the pool that would continuously cause problems every time it is used.

Yes, this is also a slight mystery to me. We _do_ have some ‘regular’ disconnect and pool timeout errors, which I initially assumed were going to be the cause of this problem, but the logs suggest otherwise: they are not closely correlated in time or count.

But this is all happening in the context of a set of Resque jobs, so perhaps the disconnect error is getting raised in a different job from the one affected by the initial problem, or only getting raised once per worker despite many of these jobs failing.

After a disconnect error, the next request automatically tries to reestablish the connection I think?

What I find odd about this aspect is that we’re seeing this particular bit of code fail _consistently_, whereas the disconnect errors are sporadic and clustered together (i.e. I think they are network-level problems, not something in the code). Could we somehow be _causing_ the initial `PG::UnableToSend`?

**Reply by jeremyevans** [reply]
> > By default, a failure to load the model schema should raise an exception, it shouldn't be silently swallowed. If it is being silently swallowed, my guess is you have set require_valid_table to false, which is a terrible idea. Alternatively, you are using a very old version of Sequel and should upgrade.
> 
> Thank you, yes that makes sense, I was confused that this didn’t raise. I can’t see `require_valid_table` being set to false anywhere in our codebase, so I don’t think it’s that. We’re on 5.34 – is that old enough to have this issue? In any case, we will upgrade. Otherwise I don’t see why this is getting swallowed, we’re certainly not rescuing anything in that part of our code. This _is_ all in DB transaction, but that only swallows `Sequel::Rollback` without re-raising I believe, and in any case, it’s only the _second_ error which causes the transaction to roll back.

require_valid_table has defaulted to true since 5.0.0 I think (https://github.com/jeremyevans/sequel/blob/5.0.0/lib/sequel/model.rb#L55).

You may want to check if the setting has been changed:

```ruby
p ModelName.require_valid_table
```

If this returns false, then something is modifying the setting.  If it returns true, then I have no idea why the error is being swallowed.

Assuming it returns false, do the following before defining your model classes:

```ruby
def (Sequel::Model).require_valid_table=(v)
  raise "problem" unless v
  super
end
```

That will hopefully alert you to where it is being set to false, assuming the code that does so does not set the instance variable directly.


> > What happens on future use of the same connection? Do you get a different error that is treated as a disconnect error? I would assume you must, otherwise you would have a disconnected connection in the pool that would continuously cause problems every time it is used.
> 
> Yes, this is also a slight mystery to me. We _do_ have some ‘regular’ disconnect and pool timeout errors, which I initially assumed were going to be the cause of this problem, but the logs suggest otherwise: they are not closely correlated in time or count.

Does this only happen in production, or can you reproduce in the testing environment?  If you can reproduce in a testing environment, try using a pool size of 1 and setting up a database logger, so you can see each query being sent, and any errors raised.  If you can only reproduce in production, try using a logger and using the `:log_connection_info` database option to include connection information in the logs.

> After a disconnect error, the next request automatically tries to reestablish the connection I think?

Assuming Sequel recognizes it as a disconnect error (Sequel::DatabaseDisconnectError), then yes, the connection pool will remove the connection, and a new connection will be established later as needed.

> What I find odd about this aspect is that we’re seeing this particular bit of code fail _consistently_, whereas the disconnect errors are sporadic and clustered together (i.e. I think they are network-level problems, not something in the code). Could we somehow be _causing_ the initial `PG::UnableToSend`?

I'm not sure how you could be causing the issue.  The underlying problem does seem likely to be network related.  It's not uncommon for network issues to occur for idle or long lived connections, and Sequel has connection_validator and connection_expiration extensions to help with those issues.  The issue you are describing sounds like you have network issues during application startup (when loading model classes), which is quite bad.  The best advice I can give you would be to try to fix the network issues so your connection to the database is reliable, but I realize that may be out of your control.

**Reply by ljfc** [reply]
> Does this only happen in production, or can you reproduce in the testing environment? If you can reproduce in a testing environment [...]

> The underlying problem does seem likely to be network related. It's not uncommon for network issues to occur for idle or long lived connections [...]

We’ve been able to recreate it in a minimal local console, connected to the production DB. It seems to be exactly what you say: the issue occurs when the connection has been idle for a bit. Possibly the ‘consistent’ failures in production are just because the Resque jobs are sporadic. That might also explain why the webapps are not affected, because they are getting hit by healthchecks which then touch the DB so maybe that’s keeping them alive.

I’ll look at those extensions and see if they help, thank you.

(Yes, we would certainly like to move the DB to the same location as everything else for a bunch of reasons, but that’s not straightforward right now.)

```
p ModelName.require_valid_table
```

This is `true`.

What’s more, in the console where we recreated the error it _does_ raise, and _does_ recover the connection so subsequent queries are fine. I’ll keep digging in Resque to try and figure out what’s different there. I also think possibly we’re initialising two connections to the database when Resque boots, could that leave Sequel unable to recover? In any case I’m going to try to remove that behaviour.

Finally, I noticed that these errors are almost always happening immediately after we run `select pg_try_advisory_lock(...)`, but I guess that’s just coincidence because we’re always doing that as we start the set of actions.

**Reply by jeremyevans** [reply]
> We’ve been able to recreate it in a minimal local console, connected to the production DB. It seems to be exactly what you say: the issue occurs when the connection has been idle for a bit. Possibly the ‘consistent’ failures in production are just because the Resque jobs are sporadic. That might also explain why the webapps are not affected, because they are getting hit by healthchecks which then touch the DB so maybe that’s keeping them alive.

You could try using the connection_validator extension, which validates connections before use.  I'm not sure how this is an issue with idle connections, unless you are not loading the related model class at application startup.

> (Yes, we would certainly like to move the DB to the same location as everything else for a bunch of reasons, but that’s not straightforward right now.)
> 
> ```
> p ModelName.require_valid_table
> ```
> 
> This is `true`.

This lends more evidence to the related class not being loaded at startup.  If you are delaying loading the class until a specific job runs, then what is probably happening is the class partially loads, but raises an exception.  Then later, some other code uses the partially loaded class.

The simplest workaround would be to load all related classes at application startup. However, if that will not work, you could try working around this issue by switching from:

```ruby
class ModelName < Sequel::Model
  # ...
end
```

to:

```ruby
ModelName = Class.new(Sequel::Model(:table_name)) do
  # ...
end
```

The difference between these two cases is if there is an exception raised while defining the model, in the first case, you are left with a partially loaded class, while in the second case, you are not, as the constant is not set until after the class is fully defined.

> What’s more, in the console where we recreated the error it _does_ raise, and _does_ recover the connection so subsequent queries are fine. I’ll keep digging in Resque to try and figure out what’s different there. I also think possibly we’re initialising two connections to the database when Resque boots, could that leave Sequel unable to recover? In any case I’m going to try to remove that behaviour.

It looks like the disconnect handling may be working at intended then.  Can you tell from the logs if the PG::UnableToSend exception raised as a Sequel::DatabaseDisconnectError, or a regular Sequel::DatabaseError?

> Finally, I noticed that these errors are almost always happening immediately after we run `select pg_try_advisory_lock(...)`, but I guess that’s just coincidence because we’re always doing that as we start the set of actions.

I would guess so as well.

**Reply by ljfc** [reply]
> This lends more evidence to the related class not being loaded at startup. If you are delaying loading the class until a specific job runs, then what is probably happening is the class partially loads, but raises an exception. Then later, some other code uses the partially loaded class.
> 
> The simplest workaround would be to load all related classes at application startup.

Ahah! I think this might be it – these models are being loaded by Zeitwerk on demand, which must be post-fork in Resque. That sounds like it fits the pattern here: this problem keeps recurring when this particular model is loaded. Possibly the way Zeitwerk is configured is also swallowing the error in some way – I’ll look into that.

Okay, I think there are a few possible solutions for us to test here (and maybe use more than one of them if we find multiple options that work):

- Load the models before fork
- Cache the table schema
- Add connection_validator and connection_expiration extensions
- Try to get the raise/catch working so we can recover

> It looks like the disconnect handling may be working at intended then. Can you tell from the logs if the PG::UnableToSend exception raised as a Sequel::DatabaseDisconnectError, or a regular Sequel::DatabaseError?

I can’t tell, it’s only logging `PG::UnableToSend: SSL SYSCALL error: EOF detected: SELECT \"pg_attribute\" ...`, there’s no mention of `Sequel::*`. Does that point to something else? I can see a few other `Sequel::Database*Error`s in the logs, which are all wrapping underlying `PG::*` exceptions. At the moment the DB logger is set up like so:

```
  Database = Sequel.connect(ENV["DATABASE_URL").tap do |db|
    db.extension :pg_array
    db.extension :pg_hstore
    db.extension :pg_json
    db.loggers << Logger.new(STDOUT).tap do |logger|
      logger.progname = 'BackgroundJobs SQL'
    end
  end
```

I’ll also try adding `log_connection_info = true` here. Thanks so much for your help with this, it’s incredibly useful.

**Reply by jeremyevans** [reply]
> > This lends more evidence to the related class not being loaded at startup. If you are delaying loading the class until a specific job runs, then what is probably happening is the class partially loads, but raises an exception. Then later, some other code uses the partially loaded class.
> > The simplest workaround would be to load all related classes at application startup.
> 
> Ahah! I think this might be it – these models are being loaded by Zeitwerk on demand, which must be post-fork in Resque. That sounds like it fits the pattern here: this problem keeps recurring when this particular model is loaded. Possibly the way Zeitwerk is configured is also swallowing the error in some way – I’ll look into that.



> Okay, I think there are a few possible solutions for us to test here (and maybe use more than one of them if we find multiple options that work):
> 
> * Load the models before fork
> * Cache the table schema
> * Add connection_validator and connection_expiration extensions
> * Try to get the raise/catch working so we can recover

Eager loading all models before processing requests seems like the simplest fix.  Might even save on memory if done pre-fork.

Caching the table schema should speed up loading model classes.  I do this in all of my production applications.

You probably want to use the connection_validator extension, since it sounds like your database connections are not reliable.

Trying to raise/catch does not seem like a good approach for this case.
 
> > It looks like the disconnect handling may be working at intended then. Can you tell from the logs if the PG::UnableToSend exception raised as a Sequel::DatabaseDisconnectError, or a regular Sequel::DatabaseError?
> 
> I can’t tell, it’s only logging `PG::UnableToSend: SSL SYSCALL error: EOF detected: SELECT \"pg_attribute\" ...`, there’s no mention of `Sequel::*`. 

If it's possible to print the exception class, exception message, and full backtrace for these exceptions, that would be useful. Hopefully that would allow me to see if changes should be made to Sequel to handle this case.

> Does that point to something else? I can see a few other `Sequel::Database*Error`s in the logs, which are all wrapping underlying `PG::*` exceptions. At the moment the DB logger is set up like so:
> 
> ```
>   Database = Sequel.connect(ENV["DATABASE_URL").tap do |db|
>     db.extension :pg_array
>     db.extension :pg_hstore
>     db.extension :pg_json
>     db.loggers << Logger.new(STDOUT).tap do |logger|
>       logger.progname = 'BackgroundJobs SQL'
>     end
>   end
> ```

This setup looks OK.

**Reply by ljfc** [reply]
> If it's possible to print the exception class, exception message, and full backtrace for these exceptions, that would be useful. Hopefully that would allow me to see if changes should be made to Sequel to handle this case.

I’ll look at getting a full backtrace from Resque, but when re-creating locally the logging is slightly different:

```
Sequel::DatabaseDisconnectError: PG::UnableToSend: SSL SYSCALL error: EOF detected
from /usr/local/bundle/gems/sequel-5.46.0/lib/sequel/adapters/postgres.rb:156:in `exec'
Caused by PG::UnableToSend: SSL SYSCALL error: EOF detected
from /usr/local/bundle/gems/sequel-5.46.0/lib/sequel/adapters/postgres.rb:156:in `exec'
```

This looks like it’s working as intended, I think, because `PG::UnableToSend` is wrapped in `Sequel::DatabaseDisconnectError` so from what you said Sequel will remove that connection from the pool.

I also realise now our console is using 5.46 whereas Resque is on 5.34, so we definitely need to upgrade/homogenise this all. Maybe this is something you already fixed, in which case apologies for not trying to upgrade before getting in touch!

**Comment by ljfc** [comment]
Hi @jeremyevans, thanks so much for your help, the changes we discussed above seem to have solved this issue.

I was able to get a full backtrace for the original exception by redefining `Sequel::Database#log_exception` method to print `Exception#full_message`.

```
Traceback (most recent call last):
        67: from console.rb:23:in `<main>'
        66: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_class.rb:188:in `start'
        65: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/repl.rb:15:in `start'
        64: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/repl.rb:38:in `start'
        63: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/input_lock.rb:78:in `with_ownership'
        62: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/input_lock.rb:61:in `__with_ownership'
        61: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/repl.rb:38:in `block in start'
        60: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/repl.rb:67:in `repl'
        59: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/repl.rb:67:in `loop'
        58: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/repl.rb:77:in `block in repl'
        57: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:259:in `eval'
        56: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:259:in `catch'
        55: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:260:in `block in eval'
        54: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:260:in `catch'
        53: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:261:in `block (2 levels) in eval'
        52: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:659:in `handle_line'
        51: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:290:in `evaluate_ruby'
        50: from /usr/local/bundle/gems/pry-0.14.1/lib/pry/pry_instance.rb:290:in `eval'
        49: from (pry):2:in `__pry__'
        48: from /usr/local/bundle/gems/mutations-0.9.1/lib/mutations/command.rb:80:in `run'
        47: from /srv/gems/choco/lib/choco/billing/charge_invoice.rb:42:in `execute'
        46: from /usr/local/lib/ruby/2.7.0/logger.rb:528:in `info'
        45: from /usr/local/lib/ruby/2.7.0/logger.rb:468:in `add'
        44: from /srv/gems/choco/lib/choco/billing/charge_invoice.rb:42:in `block in execute'
        43: from /srv/gems/choco/lib/choco/billing/charge_invoice.rb:126:in `billing_info'
        42: from /srv/gems/choco/lib/choco/billing/charge_invoice.rb:122:in `invoice'
        41: from /usr/local/bundle/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:26:in `require'
        40: from /usr/local/bundle/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:26:in `require'
        39: from /srv/toucan_box/lib/toucan_box/sequel_models/invoice.rb:5:in `<top (required)>'
        38: from /usr/local/bundle/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:34:in `require'
        37: from /usr/local/bundle/gems/zeitwerk-2.4.2/lib/zeitwerk/kernel.rb:34:in `require'
        36: from /srv/toucan_box/lib/toucan_box/sequel_models/billing_info.rb:5:in `<top (required)>'
        35: from /srv/toucan_box/lib/toucan_box/sequel_models/billing_info.rb:6:in `<module:ToucanBox>'
        34: from /srv/toucan_box/lib/toucan_box/sequel_models/billing_info.rb:7:in `<module:SequelModels>'
        33: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:140:in `block in def_Model'
        32: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:192:in `Model'
        31: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:597:in `set_dataset'
        30: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:765:in `get_db_schema'
        29: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:678:in `check_non_connection_error'
        28: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/database/transactions.rb:195:in `transaction'
        27: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/database/connecting.rb:270:in `synchronize'
        26: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/connection_pool/threaded.rb:92:in `hold'
        25: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/database/transactions.rb:201:in `block in transaction'
        24: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:678:in `block in check_non_connection_error'
        23: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/model/base.rb:765:in `block in get_db_schema'
        22: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/database/query.rb:159:in `schema'
        21: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/shared/postgres.rb:1433:in `schema_parse_table'
        20: from /usr/local/bundle/gems/sequel_pg-1.14.0/lib/sequel_pg/sequel_pg.rb:35:in `map'
        19: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/dataset/actions.rb:444:in `map'
        18: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/dataset/actions.rb:444:in `map'
        17: from /usr/local/bundle/gems/sequel_pg-1.14.0/lib/sequel_pg/sequel_pg.rb:81:in `each'
        16: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/dataset/actions.rb:152:in `each'
        15: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:614:in `fetch_rows'
        14: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/dataset/actions.rb:1089:in `execute'
        13: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:318:in `execute'
        12: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/database/connecting.rb:270:in `synchronize'
        11: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/connection_pool/threaded.rb:88:in `hold'
        10: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:318:in `block in execute'
         9: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:522:in `check_database_errors'
         8: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:318:in `block (2 levels) in execute'
         7: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:500:in `_execute'
         6: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:144:in `execute'
         5: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:120:in `check_disconnect_errors'
         4: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:144:in `block in execute'
         3: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:156:in `execute_query'
         2: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/database/logging.rb:48:in `log_connection_yield'
         1: from /usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:156:in `block in execute_query'
/usr/local/bundle/gems/sequel-5.34.0/lib/sequel/adapters/postgres.rb:156:in `exec': SSL SYSCALL error: EOF detected (PG::UnableToSend)
```

The lines in our application code here don’t look significant. They are:

48: Entrypoint for the Resque job
44-42: Resque job executing
39: Sequel::Model being loaded (by Zeitwerk)
36-34: Related Sequel::Model being loaded (by Zeitwerk)

Backtrace line 34 is the definition of the class which blows up later on: `class BillingInfo < Sequel::Model(:billing_infos)`.

The Zeitwerk lines in between our application code lines are in [Zeitwerk’s override of `Kernel.require`](https://github.com/fxn/zeitwerk/blob/v2.4.2/lib/zeitwerk/kernel.rb). Zeitwerk doesn’t appear to rescue exceptions here, so I’m still at a bit of a loss as to why this exception isn’t halting execution.

**Comment by jeremyevans** [comment]
To see if this problem is somewhere in Sequel, can you override `Sequel::Model.def_Model`, and have it rescue, print the exception, and reraise?  If `Sequel::Database#log_exception` is logging the exception, but `Sequel::Model.def_Model` is not, then you know the problem is in Sequel.

**Comment by ljfc** [comment]
Overriding `def_Model` like so:

```
      def def_Model(mod)                                                                                                                                                                                  
        puts "def_Model"                                                                                                                                                                         
        model = self                                                                                                                                                                                      
        mod.define_singleton_method(:Model) do |source|                                                                                                                                                   
          puts ":Model block"                                                                                                                                                                    
          begin                                                                                                                                                                                           
            model.Model(source)                                                                                                                                                                           
          rescue => e                                                                                                                                                                                     
            puts "RESCUE IN :Model block"                                                                                                                                                                 
            e.full_message                                                                                                                                                                                
            raise e                                                                                                                                                                                       
          end                                                                                                                                                                                             
        end                                                                                                                                                                                               
      rescue => e                                                                                                                                                                                         
        puts "RESCUE IN def_Model"                                                                                                                                                                        
        e.full_message                                                                                                                                                                                    
        raise e                                                                                                                                                                                           
      end
```

I see `def_Model` during app startup and then `:Model block` when the model tries to load. Then the `log_exception` message is printed, but `RESCUE IN :Model block` does not appear.

**Reply by jeremyevans** [reply]
OK, I looked more into this and Sequel is swallowing this error.  When creating a model class, Sequel runs two queries:

1. Determine columns
2. Get schema

Exceptions when running the query to get the columns are not swallowed, because that is used to determine whether the table is valid.  Exceptions when running the query to get the schema are swallowed, because there are cases where the schema cannot be determined (model based on view, set returning function, custom dataset, etc.).  For some reason, you have only the second query failing and not the first.

Maybe we can add a more strict mode, where for simple datasets, it will not swallow the exceptions.  We could also not swallow database disconnect errors, just as the code already does not swallow connection errors.  I think either change could potentially cause backwards compatibility issues, though, so I'll have to consider whether the changes are worth it.  Potentially, this could be made opt-in, but that has its own issues in that few people are likely to opt-in to the new behavior.

**Reply by ljfc** [reply]
In this case raising would have been good, yes, but I take your point about backwards compatibility, and I’ve no experience with models where the schema cannot be determined.

Perhaps a compromise would be to log a more explicit warning? Then it wouldn’t actually break any code, and there could be an option to suppress the warning if the model is known to be good without a schema. That might encourage people (a little) to either suppress the warning or turn on strict mode depending on the context.

It does seem strange that the second query is failing and not the first. The connection seems to _always_ initially fail after ~10m idle. Is it possible that Sequel would load the columns during app startup, and then they would have already been cached by the time the schema is queried? Otherwise I can’t see how this would happen either. A couple of other Sequel::Models do get loaded during startup, but they have no relations to the Model which fails to load.

**Reply by jeremyevans** [reply]
Logging a warning does seem like a reasonable compromise.  Maybe I can use that approach, and switch to an error in Sequel 6 if nobody complains.

Note that it may not be too surprising that the second query is failing and the first isn't, because the same connection isn't necessarily used for both queries.  It's also possible the there is no query to get the columns, because the result is already cached.

**Reply by jeremyevans** [reply]
I looked into this more.  It turns out the schema query is actually issued first, which is why it's more likely to hit the disconnect error.  The query to get the columns may not even be issued if the query to get the schema returns as expected.  If the query to get the schema fails due to a disconnect error, Sequel will issue a query to get the columns, but it won't try parsing the schema again, causing this error.  I think that behavior should be changed, not simply warned.  I'll work on this today and see if I can fix this issue with minimal changes to backwards compatibility.

**Reply by jeremyevans** [reply]
I pushed a couple changes. One will treat disconnect errors like connection errors.  Another is the addition of a `require_valid_schema` plugin, which can be used to ensure that models where Sequel expects to be able to parse the schema have the schema actually be parsed (this is a plugin and not by default due to backwards compatibility).

**Reply by ljfc** [reply]
Awesome! We’ll incorporate `require_valid_schema` so we’re covering all possibilities.
